solr/core/src/java/org/apache/solr/update/processor/ParseDateFieldUpdateProcessorFactory.java

```diff
@@ -172,4 +181,34 @@ public void init(NamedList args) {
       return (null == type) || type instanceof DateValueFieldType;
     };
   }
+
+  public static void validateFormatter(DateTimeFormatter formatter) {
+    // check it's valid via round-trip
+    try {
+      parseInstant(formatter, formatter.format(Instant.now()));
+    } catch (Exception e) {
+      throw new SolrException(SolrException.ErrorCode.SERVER_ERROR,
+          "Bad or unsupported pattern: " + formatter.toFormat().toString(), e);
+    }
+  }
+
+  private static Instant parseInstant(DateTimeFormatter formatter, String dateStr) {
+    final TemporalAccessor temporalAccessor = formatter.parse(dateStr);
+    // parsed successfully.  But is it a full instant or just to the day?
+    if (temporalAccessor.isSupported(ChronoField.INSTANT_SECONDS)) { // has time
+      // has offset time
+      if(temporalAccessor.isSupported(ChronoField.OFFSET_SECONDS)) {
```

Comment:
This part, checking OFFSET_SECONDS ought not to be present here, I think.  INSTANT_SECONDS should be self-sufficient.  See java.time.Parsed.resolveInstant() which gives clues as to what's going on.  The override "zone" we set earlier on the formatter is taking precedence over the OFFSET_SECONDS, leading you to try this as a work-around.  I wonder if this is a JDK bug; IMO it ought to be flipped.  

I think we may need to be smarter about when to set the default zone on the formatter and when not to.  If the format specifies the zone or has a 'Z' literal, then I think we don't; otherwise we set it.  Or we don't set it at all there and incorporate it here in parseInstant() which may be safer.  I'll propose something more concrete after I get some sleep; I have some WIP code.

Reply:
I'm feeling pretty confident about the following:
```
private static Instant parseInstant(DateTimeFormatter formatter, String dateStr) {
    final TemporalAccessor temporal = formatter.parse(dateStr);
    // Get Date; mandatory
    LocalDate date = temporal.query(TemporalQueries.localDate());//mandatory
    if (date == null) {
      throw new SolrException(SolrException.ErrorCode.SERVER_ERROR,
          "Date (year, month, day) is mandatory: " + formatter.toFormat().toString());
    }
    // Get Time; optional
    LocalTime time = temporal.query(TemporalQueries.localTime());
    if (time == null) {
      time = LocalTime.MIN;
    }

    final LocalDateTime localDateTime = LocalDateTime.of(date, time);

    // Get Zone Offset; optional
    ZoneOffset offset = temporal.query(TemporalQueries.offset());
    if (offset == null) {
      // no Zone offset; get Zone ID
      ZoneId zoneId = temporal.query(TemporalQueries.zone());
      if (zoneId == null) {
        zoneId = formatter.getZone();
        if (zoneId == null) {
          zoneId = ZoneOffset.UTC;
        }
      }
      return localDateTime.atZone(zoneId).toInstant();
    } else {
      return localDateTime.toInstant(offset);
    }
  }
```

This dispenses with trying to use INSTANT_SECONDS which would be a short-cut/optimization but we can't trust that due to my previous comment.
What's needed in the tests is an input string that actually provides an interesting timezone and/or offset, plus a default timezone of something other than UTC to show that we ignore that and use the zone or offset provided in the value.  I don't see any test that does that; the tests all provide values of referring to `Z` or zeroes (equivalent) -- which doesn't exercise the defaulting logic.  An example value taken from SOLR-12561's patch is: `Fri Oct 7 05:14:15 AKDT 2005`  (Alaska).

Reply:
I worked on test program demonstrating the JDK issue and I see now that this has already been reported:  https://bugs.java.com/bugdatabase/view_bug.do?bug_id=8177021    (we should put a comment about this in the code).  It's nice to see it has been fixed in JDK 9 (and my test program revealed that too).

During my exploration of this, I discovered that a lowercase 'z' in the pattern can parse various formats, and so can "VV".  I replaced the 5-'Z' patterns in the test file with a single lower 'z' and the tests passed.  FYI this part of the javadocs is helpful: https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatterBuilder.html#appendPattern-java.lang.String-  (builder has more detail than the DateTimeFormatter).

Reply:
Oh wow can't believe I missed that. This seems so much more robust then the previous version.